File: Python/Archive/1424.diagonal_traverse_II.py
# from collections import defaultdict

class Solution:
    def findDiagonalOrder(self, nums: List[List[int]]) -> List[int]:
        # draw examples to help find a pattern

        # Step through the grid in normal NM fashion. For every cell (r,c),
        # store number in the cell into dict with key (r+c) representing the
        # diagonal the number belongs to. The resulting dict thus has a
        # collection of diagonals in reverse order of desired, so for every key
        # flip the list of numbers and extend to result list.

        # O(MN) time and space

        rows, d = len(nums), {} # defaultdict(list)
        for r in range(rows):
            for c in range(len(nums[r])):
                diag = r + c
                if diag not in d:
                    d[diag] = []
                d[diag].append( nums[r][c] )

        res = []
        for i in range(len(d)):
            res.extend( reversed(d[i]) )
        return res

File: C++/Array/1424. Diagonal Traverse II.cpp
class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& nums) {
        vector<vector<int>> store; 
        vector<int> res; 
        
        for (int i = 0; i < nums.size(); i++)
            for (int j = 0; j < nums[i].size(); j++)
                if (store.size() <= i + j)
                    store.push_back({nums[i][j]});
                else
                    store[i + j].push_back(nums[i][j]);
        
        for (int i = 0; i < store.size(); i++)
            for (int j = store[i].size() - 1; j >= 0; j--)
                res.push_back(store[i][j]);
        
        return res; 
    }
};